python/GafferSceneUI/ShaderUI.py

```diff
@@ -364,3 +368,378 @@ def __graphEditorPlugContextMenu( graphEditor, plug, menuDefinition ) :
+		self.__pathListingWidget = GafferUI.PathListingWidget(
+			tmpPath,
+			columns = (
+				_DuplicateIconColumn( "Name", "name" ),
+				GafferUI.PathListingWidget.StandardColumn( "Type", "shader:type" ),
+				GafferUI.PathListingWidget.StandardColumn( "Value", "shader:value" ),
+				_LinkIconColumn( "Input", "shader:input" ),
+			),
+			allowMultipleSelection = True,
+			displayMode = GafferUI.PathListingWidget.DisplayMode.Tree,
+			sortable = False
+		)
+
+		self.__inputNavigateColumn = self.__pathListingWidget.getColumns()[3]
+
+		self._setWidget( self.__pathListingWidget )
+
+		self.__pathListingWidget.selectionChangedSignal().connect( Gaffer.WeakMethod( self.__updateButtonState ), scoped = False )
+		self.__pathListingWidget.buttonReleaseSignal().connect( 0, Gaffer.WeakMethod( self.__buttonRelease ), scoped = False )
+
+		self._addButton( "Cancel" )
+		self.__confirmButton = self._addButton( "OK" )
+		self.__confirmButton.clickedSignal().connect( Gaffer.WeakMethod( self.__buttonClicked ), scoped = False )
+
+		self.__parametersSelectedSignal = Gaffer.Signal1()
+
+		self.__updateButtonState()
+
+		self.setPath( _ShaderPath( self.__shaderNetworks, path = "/" ) )
+
+
+	def getPath( self ) :
+
+		return self.__path
+
+	def setPath(self, path ) :
+
+		if path.isSame( self.__path ) :
+			return
+
+		self.__path = path
+
+		self.__pathListingWidget.setPath( self.__path )
+
+	def parametersSelectedSignal( self ) :
+
+		return self.__parametersSelectedSignal
+
+	# Causes the dialogue to enter a modal state, returning the selected shader parameters
+	# once they have been selected by the user. Returns None if the dialogue is cancelled.
+	# Parameters are returned as a list of tuples of the form
+	# [ ( "shaderName", "parameterName" ), ... ]
+	def waitForParameters( self, **kw ) :
+
+		if len( self.__path.children() ) == 0 :
+			dialogue = GafferUI.ConfirmationDialogue(
+				"Shader Browser",
+				"No shaders to browse."
+			)
+			dialogue.waitForConfirmation( parentWindow = self.__parentWindow )
+
+		else :
+			button = self.waitForButton( **kw )
+
+			if button is self.__confirmButton :
+				return self.__result()
+
+		return None
+
+	def __buttonClicked( self, button ) :
+
+		if button is self.__confirmButton :
+			self.parametersSelectedSignal()( self.__result() )
+
+	def __buttonRelease( self, pathListing, event ) :
+
+		if event.button != event.Buttons.Left :
+			return False
+
+		path = pathListing.pathAt( event.line.p0 )
+		if path is None :
+			return False
+
+		column = pathListing.columnAt( event.line.p0 )
+		if column == self.__inputNavigateColumn :
+			inputPath = path.parent().parent()
+			inputParameter = path.property( "shader:input" )
+
+			if inputParameter is None or inputParameter == "---" :
+				return False
+
+			inputPath.append( inputParameter )
+
+			if inputPath.isValid() :
+
+				self.__pathListingWidget.setPathExpanded( inputPath, True )
+				self.__pathListingWidget.scrollToPath( inputPath )
```

Comment:
> If I recall correctly you tried that but had trouble because the PathListingWidget's own selection handlers were fighting you?

That was when I was using double click to jump, but it was working fine with single click as we're doing now.

I had been using `setSelection` but changed it at the last minute - happy for your input. The opposing forces are :
1. Jump to selection and highlight the shader to make it easy to locate visually, as you point out. But at the expense of losing the current selection, which may be a multiselection.
2. Jump but don't change selection. Harder to spot where you jumped to, but keeps the multiselection (that is, once I figure out how to maintain it).

I was leaning towards #2, but not sure if there's a precedent elsewhere in Gaffer or what a user may expect.

Reply:
Thanks for the explanation - makes sense. I think I still prefer option 1, because I don't think my current selection is going to be important to me when I'm still trying to navigate through connections. We should check with other folks to see if there's a consensus though - maybe I'm the outlier.

Reply:
Went with option 1 in 695714a4f121f430e45d5cb098c841426429d6a2